temporalio/bridge/src/worker.rs

```diff
@@ -79,6 +101,270 @@ pub struct ResourceBasedSlotSupplier {
+                        "Unexpected error in custom slot supplier `reserve_slot`: {}",
+                        e
+                    );
+                    continue;
+                }
+            }
+            .await;
+            match pypermit {
+                Ok(p) => {
+                    return SlotSupplierPermit::with_user_data(p);
+                }
+                Err(_) => {
+                    // This is a user thrown error, re-raised by the logging wrapper so we can
+                    // loop, so do that.
+                }
+            }
+        }
+    }
+
+    fn try_reserve_slot(&self, ctx: &dyn SlotReservationContext) -> Option<SlotSupplierPermit> {
+        Python::with_gil(|py| {
+            let py_obj = self.inner.as_ref(py);
+            let pa = py_obj.call_method1(
+                "try_reserve_slot",
+                (SlotReserveCtx::from_ctx(SK::kind(), ctx),),
+            )?;
+
+            if pa.is_none() {
+                return Ok(None);
+            }
+            PyResult::Ok(Some(SlotSupplierPermit::with_user_data(pa.into_py(py))))
+        })
+        .unwrap_or_else(|e| {
+            error!(
+                "Uncaught error in custom slot supplier `try_reserve_slot`: {}",
+                e
+            );
+            None
+        })
+    }
+
+    fn mark_slot_used(&self, ctx: &dyn SlotMarkUsedContext<SlotKind = Self::SlotKind>) {
+        if let Err(e) = Python::with_gil(|py| {
+            let permit = ctx
+                .permit()
+                .user_data::<PyObject>()
+                .cloned()
+                .unwrap_or_else(|| py.None());
+            let py_obj = self.inner.as_ref(py);
+            py_obj.call_method1(
+                "mark_slot_used",
+                (SlotMarkUsedCtx {
+                    slot_info: slot_info_to_py_obj(py, ctx.info().downcast()),
+                    permit,
+                },),
+            )?;
+            PyResult::Ok(())
+        }) {
+            error!(
+                "Uncaught error in custom slot supplier `mark_slot_used`: {}",
+                e
+            );
+        }
+    }
+
+    fn release_slot(&self, ctx: &dyn SlotReleaseContext<SlotKind = Self::SlotKind>) {
+        if let Err(e) = Python::with_gil(|py| {
+            let permit = ctx
+                .permit()
+                .user_data::<PyObject>()
+                .cloned()
+                .unwrap_or_else(|| py.None());
+            let py_obj = self.inner.as_ref(py);
+            py_obj.call_method1(
+                "release_slot",
+                (SlotReleaseCtx {
+                    slot_info: ctx.info().map(|i| slot_info_to_py_obj(py, i.downcast())),
+                    permit,
+                },),
+            )?;
+            PyResult::Ok(())
+        }) {
+            error!(
+                "Uncaught error in custom slot supplier `release_slot`: {}",
+                e
+            );
+        }
+    }
+
+    fn available_slots(&self) -> Option<usize> {
```

Comment:
I assume there is no value in allowing users to define this, at least not at this time.